runtime-modules/content/src/lib.rs

```diff
@@ -1261,6 +944,548 @@ decl_module! {
+            // Can not fail, needed check made
+            T::Currency::reserve(&participant_account_id, bid)?;
+
+            // Make auction bid & update auction data
+            let mut video = video;
+
+            if let Some(auction) = video.get_nft_auction_ref_mut() {
+                auction.make_bid(participant_id, bid, current_block);
+
+                VideoById::<T>::insert(video_id, video);
+
+                // Trigger event
+                Self::deposit_event(RawEvent::AuctionBidMade(participant_id, video_id, bid));
+            }
+        }
+
+        /// Cancel open auction bid
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn cancel_open_auction_bid(
+            origin,
+            participant_id: T::MemberId,
+            video_id: T::VideoId,
+        ) {
+
+            // Authorize participant under given member id
+            let participant_account_id = ensure_signed(origin)?;
+            ensure_member_auth_success::<T>(&participant_id, &participant_account_id)?;
+
+            // Ensure given video exists
+            let video = Self::ensure_video_exists(&video_id)?;
+
+            // Ensure auction for given video id exists
+            let auction = video.ensure_nft_auction_state::<T>()?;
+
+            // Ensure nft auction not expired
+            Self::ensure_nft_auction_not_expired(&auction)?;
+
+            let current_block = <frame_system::Module<T>>::block_number();
+
+            // Ensure auction have been already started
+            auction.ensure_auction_started::<T>(current_block)?;
+
+            // Ensure participant can cancel last bid
+            auction.ensure_bid_can_be_canceled::<T>(participant_id, current_block)?;
+
+            //
+            // == MUTATION SAFE ==
+            //
+
+            // Cancel last auction bid & update auction data
+            let mut video = video;
+
+            if let Some(auction) = video.get_nft_auction_ref_mut() {
+                auction.cancel_bid();
+
+                VideoById::<T>::insert(video_id, video);
+
+                // Trigger event
+                Self::deposit_event(RawEvent::AuctionBidCanceled(participant_id, video_id));
+            }
+        }
+
+        /// Complete auction
+        #[weight = 10_000_000] // TODO: adjust weight
+        pub fn complete_video_auction(
+            origin,
+            member_id: T::MemberId,
+            video_id: T::VideoId,
+            metadata: Metadata,
+        ) {
+
+            // Ensure given video exists
+            let video = Self::ensure_video_exists(&video_id)?;
+
+            // Ensure auction for given video id exists, retrieve corresponding one
+            let auction = video.ensure_nft_auction_state::<T>()?;
+
+            let last_bid = auction.ensure_last_bid_exists::<T>()?;
+
+            // Ensure actor authorized to complete auction.
+            Self::ensure_actor_is_last_bidder(origin, member_id, &auction)?;
```

Comment:
I disagree with you, because the prior `ensure_last_bid_exists` check guarantee last bid exists. 
Only the bidder can complete auction, if there is no bidder present, auction should be cancelled, not completed!

Reply:
If there is no bidder, then `complete_nft_auction` should not give `Ok`, we shuold keep things separate, if you want to cleanup an auction with any bidders, then you should do `cancel_auction`, not `complete`.

Reply:
I think there is some misunderstanding present. 
In first message you are saying 
>This codepath can fail statically, in auction.ensure_caller_is_last_bidder, because there is no bid, while this is impossible due to ensure on prior line.
>
but 
```rust
let last_bid = auction.ensure_last_bid_exists::<T>()?;
```
guarantee bid present. If there is no bidder, `then complete_nft_auction` extrinsic will return an error. 
We should cancel auctin in that case.

>If there is no bidder, then complete_nft_auction should not give Ok, we shuold keep things separate, if you want to cleanup an auction with any bidders, then you should do cancel_auction, not complete
>
Exactly, the current logic flow of this extrinsics is like you are describing here.